### ISSUE TITLE
Use only long args and use --deps?

### ISSUE THREAD

**Issue by david-a-wheeler** [boyski/pmaudit#5]
> Maybe this should be parse_known_args() with all unrecognized args passed on to the shell?

If you seriously want unknown options to be a pass-through to a shell, then I suggest using only long options. Currently -j and -d are not used by common shells, but there is no guarantee for that in the future. Long options are much less likely to be duplicated, so using only long options would be far more future proof.

And if options are only long, then I suggest --deps instead of --depsfile. The extra characters don't really add anything, except making the command itself longer. Which wouldn't matter so much if you can just use a short single letter, but does in this case.

**Comment by boyski** [comment]
> If you seriously want unknown options to be a pass-through to a shell ...

I do think it's worth seriously considering but only when -c is used. I
would imagine passing them through in that case or rejecting them with
"unknown option" otherwise. I know this is easy with argparse. One question
is how easy it is to do with getopt_long().

> Currently -j and -d are not used by common shells, but there is no
guarantee for that in the future.

My view on this has been to split the difference. For one thing, I don't
think the possibility that bash would add a -d or -j is a huge concern
because (a) it's a mature product at this point, (b) its maintainer(s)
would also have to be aware of the downstream impact of disturbing its CLI,
and (c) if such a flag was added the likelihood of it being something that
might get into .SHELLFLAGS is pretty low, considering that in all these
years make only uses -c and sometimes -e, and even my overrides only get as
far as -u and some -o options. Anything added now is likely to be subtle or
special-case enough that it wouldn't matter to make.

But the second point is the more important: by providing both short flags
and long we put the future-proofing decision in the hands of users. My own
mnemonic is to use long flags in scripted uses since I won't be retyping
them often and the result is easier for future maintainers or logfile
analyzers to read but use short options manually because I'm a bad typist.
Bottom line, as long as a unique long option is provided the user always
has a solution to compatibility issues and whether they want it is up to
them.

> I suggest --deps instead of --depsfile. The extra characters don't really
add anything.

I tend to go the other way on this because argparse does auto-abbreviation
so for long flags I prefer fully descriptive names which can then be
abbreviated. In other words we can use --depsfile and opts.depfile within
the code while the user can use --deps. Again, I do not remember whether
getopt_long() does the same.

David

**Comment by david-a-wheeler** [comment]
Happy to support short flagas well (when they don't conflict with current shell flags), and happy to use --depsfile instead of --deps.

For now, I think we should try to directly support and pass through shell flags, instead of just arbitrarily supporting anything unknown. If we just pass anything then we lose immediate error detection since it won't happen till later (at that may be confusing). There is also the problem of being uncertain of what exactly the pass, for example, if an unknown option has a parameter and the parameter happens to also start with a dash. I expect that a relatively small number of flags will be all that anyone expects the pass anyway, and we can handle those directly. People can always create intermediate shells to handle more complicated cases, so no serious functionality is lost. And we can always add more general flag handling later, but it is hard to disable it once users start depending on it.

**Comment by boyski** [comment]
Fine with me.

**Comment by boyski** [comment]
BTW the manual for getopt_long() says "Long option names may be abbreviated
if the abbreviation is unique ..." so no problem there.

**Comment by david-a-wheeler** [comment]
Okay, we obviously will *not* use just long args, but we'll be judicious about which options have short option names.  Resolved, so closing.
